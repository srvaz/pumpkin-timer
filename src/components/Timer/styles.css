.timer-container {
  align-items: center;
  background-color: var(--primary-color-light);
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  margin-top: 100px;
  padding: var(--space-st) var(--space-xl);
}

.timer-container .timer-count {
  color: var(--primary-color-dark);
  font-size: var(--font-size-xxl);
}

.timer-container button {
  background-color: var(--primary-color-dark);
  border-radius: 4px;
  border: none;
  color: var(--text-color-light);
  font-size: var(--font-size-lg);
  font-weight: var(--font-weight-bold);
  margin-top: var(--space-st);
  padding: var(--space-md);
  text-transform: uppercase;
  width: 100%;
}

.timer-container button:hover {
  cursor: pointer;
  opacity: 0.85;
}
