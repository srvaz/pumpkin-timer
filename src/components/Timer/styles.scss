.timer-card {
  height: 200px;
  position: relative;

  .timer-progress {
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
  }

  .timer-time {
    position: relative;
  }
}
