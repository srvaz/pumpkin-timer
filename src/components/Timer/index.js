import React, { Component } from 'react';
import { Card, CardContent, IconButton, Typography } from '@material-ui/core';
import {
  PlayArrow as PlayArrowIcon,
  Stop as StopIcon,
} from '@material-ui/icons';

import './styles.css';
import { normalizeNumber, dateToTime } from '../../utils/numbers';

export default class Timer extends Component {
  state = {
    minutes: 0,
    seconds: 0,
    started: false,
  };

  timerInterval = null;

  componentDidMount() {
    this.resetTime();
  }

  resetTime = () => {
    const now = new Date();
    const startTime = now.getTime();
    const endTime = now.setMinutes(
      now.getMinutes() + parseInt(this.props.minutes)
    );

    const { minutes, seconds } = dateToTime(startTime, endTime);

    this.setState({ minutes, seconds });
  };

  toggleTimer = () => {
    const { started } = this.state;

    started ? this.stopTimer() : this.startTimer();

    this.setState({ started: !started });
  };

  countDown = (countDownEnd) => {
    this.timerInterval = setInterval(() => {
      const now = new Date().getTime();

      const { distance, minutes, seconds } = dateToTime(now, countDownEnd);

      if (distance >= 0) {
        this.setState({ minutes, seconds });
      } else {
        this.resetTime();
        clearInterval(this.timerInterval);
      }
    }, 1000);
  };

  startTimer = () => {
    const countDownEnd = new Date();
    countDownEnd.setMinutes(
      countDownEnd.getMinutes() + parseInt(this.props.minutes)
    );

    this.countDown(countDownEnd.getTime());
  };

  stopTimer = () => {
    clearInterval(this.timerInterval);
    this.resetTime();
  };

  render() {
    const { minutes, seconds, started } = this.state;

    return (
      <Card className='timer-card' variant='outlined'>
        <CardContent>
          <Typography variant='h5'>
            {normalizeNumber(minutes)}:{normalizeNumber(seconds)}
          </Typography>
          <IconButton
            id='js-icon-button'
            onClick={this.toggleTimer}
            aria-label='Play/Stop'
          >
            {started ? <StopIcon /> : <PlayArrowIcon />}
          </IconButton>
        </CardContent>
      </Card>
    );
  }
}
