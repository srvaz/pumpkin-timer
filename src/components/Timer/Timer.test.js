import React from 'react';
import { render } from '@testing-library/react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import Timer from './';

configure({ adapter: new Adapter() });

describe('Components > Timer', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <Timer minutes='1' />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('countdown triggers', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<Timer minutes='1' />);
    });

    test('toogle timer', async () => {
      const button = wrapper.find('#js-icon-button');

      // Starts the countdown
      button.simulate('click');
      expect(wrapper.state('started')).toBe(true);

      // Stops the countdown
      button.simulate('click');
      expect(wrapper.state('started')).toBe(false);
    });
  });
});
