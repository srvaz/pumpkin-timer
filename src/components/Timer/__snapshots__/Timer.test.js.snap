// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > Timer renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root timer-card MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        01:00
      </h5>
      <button
        aria-label="Play/Stop"
        class="MuiButtonBase-root MuiIconButton-root"
        id="js-icon-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8 5v14l11-7z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
